daemon/src/main/java/org/apache/maven/cli/DaemonMavenCli.java

```diff
@@ -181,1156 +166,1194 @@ public int doMain(CliRequest cliRequest) throws Exception
-        {
-            if ( String.valueOf( CLIManager.SET_SYSTEM_PROPERTY ).equals( opt.getOpt() ) )
-            {
-                setPropertyOptions.add( opt );
-            }
-            else
-            {
-                commandLineBuilder.addOption( opt );
-            }
-        }
-        for ( Option opt : mavenConfig.getOptions() )
-        {
-            commandLineBuilder.addOption( opt );
-        }
-        // finally add the CLI system properties
-        for ( Option opt : setPropertyOptions )
-        {
-            commandLineBuilder.addOption( opt );
-        }
-        return commandLineBuilder.build();
-    }
+            modelProcessor = container.lookup( ModelProcessor.class );
 
-    /**
-     * configure logging
-     */
-    void logging( CliRequest cliRequest )
-    {
-        // LOG LEVEL
-        cliRequest.debug = cliRequest.commandLine.hasOption( CLIManager.DEBUG );
-        cliRequest.quiet = !cliRequest.debug && cliRequest.commandLine.hasOption( CLIManager.QUIET );
-        cliRequest.showErrors = cliRequest.debug || cliRequest.commandLine.hasOption( CLIManager.ERRORS );
+            configurationProcessors = container.lookupMap( ConfigurationProcessor.class );
 
-        java.util.logging.Level level;
-        if ( cliRequest.debug )
-        {
-            cliRequest.request.setLoggingLevel( MavenExecutionRequest.LOGGING_LEVEL_DEBUG );
-            level = Level.FINER;
-        }
-        else if ( cliRequest.quiet )
-        {
-            cliRequest.request.setLoggingLevel( MavenExecutionRequest.LOGGING_LEVEL_ERROR );
-            level = Level.SEVERE;
-        }
-        else
-        {
-            cliRequest.request.setLoggingLevel( MavenExecutionRequest.LOGGING_LEVEL_INFO );
-            level = Level.INFO;
-        }
+            toolchainsBuilder = container.lookup( ToolchainsBuilder.class );
 
-        // LOG COLOR
-        String styleColor = cliRequest.getUserProperties().getProperty( STYLE_COLOR_PROPERTY, "auto" );
-        if ( "always".equals( styleColor ) )
-        {
-            MessageUtils.setColorEnabled( true );
-        }
-        else if ( "never".equals( styleColor ) )
-        {
-            MessageUtils.setColorEnabled( false );
-        }
-        else if ( !"auto".equals( styleColor ) )
-        {
-            throw new IllegalArgumentException( "Invalid color configuration option [" + styleColor
-                + "]. Supported values are (auto|always|never)." );
+            dispatcher = (DefaultSecDispatcher) container.lookup( SecDispatcher.class, "maven" );
         }
-        else if ( cliRequest.commandLine.hasOption( CLIManager.BATCH_MODE )
-            || cliRequest.commandLine.hasOption( CLIManager.LOG_FILE ) )
-        {
-            MessageUtils.setColorEnabled( false );
+
+        @Override
+        public void close() {
+            eventSpyDispatcher.close();
+            container.dispose();
+            purgeRealms();
```

Comment:
This is the main idea of the fix: we do not re-use the container (an all its dependents) across mvnd executions and we remove -SNAPSHOTS and in-source-tree jars from the class world.

The fact that we do not ditch the whole classworld is intended to keep as many classes as possible loaded and warm. We throw away only those class loaders that are known to be mutable. We still assume that the rest of the jars in the local Maven repo is immutable. https://github.com/mvndaemon/mvnd/issues/26 is an example of a situation where it does not hold and we should have some checks in place to prevent it.

I was pondering whether we could keep parts of the plexus container across mvnd executions too (to speedup the repeated builds). It seems quite risky.  I am not sure we would be able to reliably remove only all the stale beans and their dependents. The beans can be linked in uncontrollable ways. I have not (quickly) found a way how to do that.

Reply:
Isn't that supposed to be done by the CliPluginRealmCache already ?

Reply:
Actually yes, see #37 

BTW, I have not see CliPluginRealmCache to be picked. I had to add an explicit binding https://github.com/mvndaemon/mvnd/pull/37/files#diff-085c748cd75673826a31723d0840b82eR486